Fix cross compilation for armv6hf (Raspberry Pi 1)

starypatyk opened the pull request:
The Docker image intended to cross-compile librespot for armv6hf (i.e. Raspberry Pi 1) did not work for some time.

Changes proposed here attempt to fix this:

- ~~Commit https://github.com/librespot-org/librespot/commit/faeaf506d64ab988d6cc950b58dbe22c0eed9386 changes the default tls library from **aws-lc-rs** to **ring** as suggested by @kingosticks in #1382 (commit: https://github.com/kingosticks/librespot/commit/24bbc6314c991c64d1846162f739f223c44e5779). I tried to compile **aws-lc-rs** for armv6hf, but did not succeed. :disappointed: I understand that this change might be controversial - I am open to discuss this.~~ Reverted in https://github.com/librespot-org/librespot/pull/1457/commits/6bb009af83ac737c60c99efbf5d8598f6debd3d8
- Commit https://github.com/librespot-org/librespot/commit/0f497cc690e4fefef83686864155d6d68ab2977f adds **with-libmdns** to the list of features - to compensate for the changes made in #1347.
- Commit https://github.com/librespot-org/librespot/commit/3026d030b552e5fa4cb2c7d8f4981a7d2f069fc7 is just a small clean-up of the Docker image and the script used to build librespot.
- Commit https://github.com/librespot-org/librespot/pull/1457/commits/765f1f2246ba0cd7f8776e983a4dfffd539e4395 fixes build of the **aws-lc-rs** library.
- Commit https://github.com/librespot-org/librespot/pull/1457/commits/ab1102a0aa36d44dbca253de7819b0db844bf0f7 fixes a minor issue in binaries generated via Docker for other platforms (VERGEN_IDEMPOTENT_OUTPUT instead of actual git commit hash)

With these changes in place I have been able to cross-compile librespot for armv6hf and verify that it runs as expected on a Raspberry Pi 1.

I have also verified builds using Docker for aarch64 and x86_64.

kingosticks commented:
I've not tried to compile for pi 1 lately but I would be extremely surprised if switching to ring is a requirement for that to work. It never used to be. 

I think we should put the default back to ring. I have been meaning to get back to that PR but I really don't think it should be merged in as part of this, it's too much of a big change.

starypatyk commented:
> I've not tried to compile for pi 1 lately but I would be extremely surprised if switching to ring is a requirement for that to work. It never used to be.

After installing cmake (from system packages) and bindgen-cli (using cargo) I got to the point of missing **libclang**.
```
Unable to find libclang: "couldn't find any valid shared libraries matching: ['libclang.so', 'libclang-*.so', 'libclang.so.*', 'libclang-*.so.*'], set the `LIBCLANG_PATH` environment variable to a path where one of these files can be found (invalid: [])"
```

I have not been able to find **libclang** in the set of armv6hf packages used by the Dockerfile. An alternative might be to use e.g. these tools (https://github.com/tttapa/docker-arm-cross-toolchain) instead. But I have not tried this yet.

> I think we should put the default back to ring. I have been meaning to get back to that PR but I really don't think it should be merged in as part of this, it's too much of a big change.

Fair enough - I can remove this change from this PR and wait for another PR that will make this explicit.

kingosticks commented:
It might be missing `apt install clang libclang-dev` ? Might as well update the base image, 18.04 is ridiculous.

starypatyk commented:
> It might be missing `apt install clang libclang-dev` ?

Tried that - got a different error this time:

```
  bindgen-STDERR: /usr/include/stdint.h:26:10: fatal error: 'bits/libc-header-start.h' file not found
  Unable to generate bindings: clang diagnosed error: /usr/include/stdint.h:26:10: fatal error: 'bits/libc-header-start.h' file not found
```

I suspected that **libclang** is required for the target architecture, not for the architecture used on the build host.

> Might as well update the base image, 18.04 is ridiculous.

Indeed - I did not want to start with major changes.

kingosticks commented:
Bindgen is used to generate rust code (not binaries) from c headers, I'd have thought it would be fine with the host version. But it does need to have the target's c headers to do this task, think that might be the problem. Where are they provided? In the PI tools sysroot or from apt?

kingosticks commented:
But to be honest, you are right. It's a waste of time fighting aws-lc all the time. Let's see if we can switch to ring as default and make everyone's lives 10x easier.

starypatyk commented:
> Bindgen is used to generate rust code (not binaries) from c headers, I'd have thought it would be fine with the host version. But it does need to have the target's c system (not clang) headers to do this task, think that might be the problem. Where are they provided? In the PI tools sysroot or from apt?

@kingosticks - thanks for this tip. I have finally managed to cross-compile **aws-lc-rs**. :+1:

Setting **BINDGEN_EXTRA_CLANG_ARGS** environment variable did the trick - I have found it here: https://crates.io/crates/bindgen-cli

The compiled librespot binary seems to run on a Raspberry Pi 1.

I am not sure if it makes sense to include required changes in this PR. Maybe I will clean them up and submit - just in case. :thinking:

roderickvd commented:
It used to be ring before. Switching to aws-lc must have happened by me merging a Docker PR and not being aware of the implications. We don’t need the FIPS compliance and certainly not the hard time that it’s giving us.

starypatyk commented:
> It used to be ring before. Switching to aws-lc must have happened by me merging a Docker PR and not being aware of the implications. We don’t need the FIPS compliance and certainly not the hard time that it’s giving us.

Apparently this has been changed upstream in [hyper-rustls](https://crates.io/crates/hyper-rustls) at [0.27.0 release](https://github.com/rustls/hyper-rustls/releases/tag/v%2F0.27.0):

> Default cryptography provider changed to [aws-lc-rs](https://github.com/aws/aws-lc-rs) to match upstream Rustls project. Note that this has some implications on [platform support and build-time tool requirements](https://aws.github.io/aws-lc-rs/requirements/index.html) such as cmake on all platforms and nasm on Windows.
> Support for ring continues to be available: set the ring crate feature.

starypatyk commented:
Since I have found a way to build **aws-lc-rs** for armv6hf, I have reverted the change of the default TLS library to leave it for a separate explicit PR.

Brief summary of changes required to build **aws-lc-rs** (might help someone in the future):

- Install additional system packages **cmake**, **clang** and **libclang-dev**.
- Install bindgen (`cargo install --force --locked bindgen-cli`)
- Set `BINDGEN_EXTRA_CLANG_ARGS` environment variable to point to sysroot for relevant **target** architecture.

I have also added **git** to the packages in the `contrib/Dockerfile` to fix a minor issue - to have correct commit hash rather than VERGEN_IDEMPOTENT_OUTPUT when checking version of the resulting binary.

I do not plan any more changes to this PR - unless there are further comments/suggestions.

photovoltex commented on contrib/cross-compile-armv6hf/docker-build.sh:
```diff
@@ -1,14 +1,17 @@
+cargo install --force --locked bindgen-cli
+
+PI1_TOOLS_DIR=/pi/tools/arm-bcm2708/arm-linux-gnueabihf
+PI1_TOOLS_SYSROOT_DIR=$PI1_TOOLS_DIR/arm-linux-gnueabihf/sysroot
```
Just out of curiosity. This path extend to `/pi/tools/arm-bcm2708/arm-linux-gnueabihf/arm-linux-gnueabihf/sysroot`. Is it intentional that `arm-linux-gnueabihf` is repeated in the path? It seemed to be previously the same, but it feels a bit off, so I just wanted to ask if that is the expected behavior :)

starypatyk replied:
Yes - this is how the files are arranged in the https://github.com/raspberrypi/tools repository.

Please note that `/pi/tools/arm-bcm2708/arm-linux-gnueabihf` is a symlink to `/pi/tools/arm-bcm2708/arm-rpi-4.9.3-linux-gnueabihf`, so contents of the sysroot directory comes from: https://github.com/raspberrypi/tools/tree/master/arm-bcm2708/arm-rpi-4.9.3-linux-gnueabihf/arm-linux-gnueabihf/sysroot

I must admit that I am not so fluent in cross-compilation to fully understand what is where in this structure. :disappointed:

photovoltex replied:
Ah oki, interesting. Just looks a bit like a minor mistake from the outside, but all good then^^

photovoltex commented on contrib/cross-compile-armv6hf/Dockerfile:
```diff
@@ -11,20 +11,18 @@ FROM --platform=linux/amd64 ubuntu:18.04
 
 # Install rust.
-RUN curl https://sh.rustup.rs -sSf | sh -s -- -y
+RUN curl https://sh.rustup.rs -sSf | sh -s -- --default-toolchain 1.81 -y
```
Is this required? Or what are the intentions behind this change?

starypatyk replied:
Strictly - not required. I wanted to make rust version explicit and aligned with MSRV,

photovoltex replied:
Make sense, just wanted to know if there was a deeper meaning behind it. For example missing comparability in stable or so.

photovoltex reviewed (approved):
Looks good to me, just a last small question
